spec/consensus/reactor/reactor.md

```diff
@@ -0,0 +1,333 @@
+
+| Eventual $\Delta$-Timely Communication|
+|-----|
+|There is a bound $\Delta$ and an instant GST (Global Stabilization Time) such that if a correct process $p$ sends a message $m$ at a time $t \geq \text{GST}$ to a correct process $q$, then $q$ will receive $m$ before $t + \Delta$.
+
+|Gossip communication|
+|-----|
+| (i) If a correct process $p$ sends some message $m$ at time $t$, all correct processes will receive $m$ before $\text{max} \{t,\text{GST}\} + \Delta$.
+| (ii) If a correct process $p$ receives some message $m$ at time $t$, all correct processes will receive $m$ before $\text{max}\{t,\text{GST}\} + \Delta$.
+
+Roughly speaking, if Gossip Communication is guaranteed, then all messages sent by correct processes will be eventually delivered to all correct processes and allow all correct processes to realize that the same conditions are.
+
+However, since processes are subject to failures, it is not guaranteed that proposals are ever sent and correct processes cannot wait indefinitely.
+Hence, processes execute in rounds in which they wait for conditions to be met for sometime and, if they timeout, send negative messages that will lead to new rounds.
+Again, Gossip Communication guarantees that eventually the conditions for deciding are met, even if only after GST is reached.
+
+Implementing Gossip Communication, however, is hard because even messages sent before the GST need to be buffered to be reliably delivered between correct processes, but since GST may take indefinitely long to arrive, in practice implementing this property would require unbounded memory.
+
+Fortunately, while Gossip Communication is a sufficient condition for the Tendermint algorithm to terminate, it is not strictly necessary, because the conditions to progress and terminate are evaluated over the messages of subsets of rounds executed, not all of them, and as new rounds are executed, messages in old rounds may be become obsolete and be ignored and forgotten, saving memory.
+In other words, the algorithm does not require all messages to be delivered, only messages that advance the state of the processes.
+
+## Node's state as a Tuple Space
+
+One way of looking at the information used by CometBFT nodes is as a distributed tuple space to which all nodes contribute with signed entries;
+to share a proposal, the proposer adds it to the tuple space and to vote for a proposal, the node adds the vote.
+Each update includes the author name and therefore they are all non-conflicting with each other and update forgery is prevented with signing each update.
+
+Since we are talking about an asynchronous distributed system, individual nodes can only maintain approximations of the tuple space.
+Nodes may broadcast the update operations to all nodes, including themselves, and, if the communication is reliable, as guaranteed by Gossip Communication, the tuple space will eventually converge.
```

Comment:
I'm not sure I follow how gossip guarantees that communication between nodes is reliable. 

You define a concept of GST that asserts an upper bound for "correct" processes to receive messages, but in the case of a network fault, all processes would be classified as "incorrect" — it's not clear to me how this outcome produces reliable communication?

Reply:
When I say Gossip Communication I am not talking about gossip as a general technique. I am talking about the very precise definition given in the previously in the document, which came from the Tendermint paper.
Gossip Communication, as defined, is reliable for all messages seen by correct processes.

Reply:
> Gossip Communication, as defined, is reliable for all messages seen by correct processes.

That's fine, as long as it's understood that this definition of Gossip Communication is not actually satisfied by the  current CometBFT implementation.

Reply:
Indeed it is and this is one of the reasons for this spec.

> Fortunately, while Gossip Communication is a sufficient condition for the Tendermint algorithm to terminate, it is not strictly necessary, because the conditions to progress and terminate are evaluated over the messages of subsets of rounds executed, not all of them, and as new rounds are executed, messages in old rounds may be become obsolete and be ignored and forgotten, saving memory.
> In other words, the algorithm does not require all messages to be delivered, only messages that advance the state of the processes.